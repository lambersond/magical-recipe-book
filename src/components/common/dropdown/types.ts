export type DropdownOption<S = string, T = string> = {
  id: S
  label: React.ReactNode
  value?: T
}

export type DropdownProps<S = string, T = string | number> = {
  options: DropdownOption<S, T>[]
  selected?: DropdownOption<S, T>
  onSelect: (option: DropdownOption<S, T>) => void
  label?: string
  name?: string
  required?: boolean
  defaultEmpty?: boolean
  placeholder?: string
  width?: string
}
