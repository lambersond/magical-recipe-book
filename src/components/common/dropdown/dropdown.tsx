'use client'

import { useState, useRef, useEffect } from 'react'
import { ArrowDownIcon } from '../icons/arrow-down'
import type { DropdownOption, DropdownProps } from './types'

export function Dropdown<S, T>({
  options,
  onSelect,
  width = 'w-72',
  label,
  name = 'dropdown',
  selected,
  defaultEmpty = false,
  placeholder = 'No options available',
  ...props
}: Readonly<DropdownProps<S, T>>) {
  const [open, setOpen] = useState(false)
  const [choice, setChoice] = useState(() =>
    defaultEmpty || !options?.[0]
      ? { id: 'default-empty', label: placeholder }
      : options[0],
  )
  const dropdownRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setOpen(false)
      }
    }

    document.addEventListener('mousedown', handleClickOutside)
    return () => document.removeEventListener('mousedown', handleClickOutside)
  }, [])

  useEffect(() => {
    if (selected) {
      setChoice(selected)
    }
  }, [selected])

  const handleSelect = (option: DropdownOption<S, T>) => () => {
    setChoice(option)
    onSelect(option)
    setOpen(false)
  }

  return (
    <div
      ref={dropdownRef}
      className='relative inline-flex flex-col text-left w-fit rounded-md'
    >
      {!!label && (
        <label
          className='text-sm text-text-secondary font-bold uppercase'
          htmlFor={name}
        >
          {label}
          {props.required && <sup>*</sup>}
        </label>
      )}
      <button
        name={name}
        type='button'
        data-testid='Dropdown__button'
        onClick={() => setOpen(!open)}
        className={`
          inline-flex justify-between ${width} rounded-md border border-border
          pl-4 pr-2 py-2 bg-paper text-md text-text-primary focus:outline-none
          cursor-pointer hover:bg-paper/75
        `}
      >
        {choice.label}
        <ArrowDownIcon className='ml-auto' />
      </button>
      {open && (
        <div
          className={`origin-bottom-right absolute left-0 top-full ${width} rounded-sm bg-paper shadow-lg`}
        >
          <div className='py-1 cursor-pointer'>
            {options.map(option => (
              <button
                key={`${option.id}-${option.label}`}
                data-testid={`Dropdown__option-${option.id}`}
                onClick={handleSelect(option)}
                className='w-full text-left px-4 py-2 text-md text-text-secondary hover:text-text-primary hover:bg-neutral-950/50 cursor-pointer'
              >
                {option.label}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  )
}
