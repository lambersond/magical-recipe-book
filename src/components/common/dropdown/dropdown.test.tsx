import { render, useClick, screen } from '@test-utils'
import { Dropdown } from './dropdown'

describe('components/common/dropdown', () => {
  const options = [
    { id: '1', value: '1', label: 'Option 1' },
    { id: '2', value: '2', label: 'Option 2' },
    { id: '3', value: '3', label: 'Option 3' },
  ]

  it('should render correctly', () => {
    expect(
      render(
        <Dropdown options={options} label='test' onSelect={jest.fn()} />,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render correctly with defaultEmpty', () => {
    expect(
      render(
        <Dropdown
          options={options}
          label='test'
          onSelect={jest.fn()}
          defaultEmpty
        />,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render options on click', async () => {
    const click = useClick(<Dropdown options={options} onSelect={jest.fn()} />)

    await click(screen.getByText('Option 1'))

    expect(screen.getByTestId('Dropdown__option-1')).toHaveTextContent(
      'Option 1',
    )
    expect(screen.getByTestId('Dropdown__option-2')).toHaveTextContent(
      'Option 2',
    )
    expect(screen.getByTestId('Dropdown__option-3')).toHaveTextContent(
      'Option 3',
    )
  })

  it('should call onSelect on option click', async () => {
    const onSelect = jest.fn()
    const click = useClick(<Dropdown options={options} onSelect={onSelect} />)

    await click(screen.getByText('Option 1'))

    expect(onSelect).toHaveBeenCalledTimes(0)

    await click(screen.getByTestId('Dropdown__option-2'))

    expect(onSelect).toHaveBeenCalledWith(options[1])
  })

  it('should close dropdown on click outside', async () => {
    const click = useClick(<Dropdown options={options} onSelect={jest.fn()} />)

    await click(screen.getByText('Option 1'))

    expect(screen.getByTestId('Dropdown__option-1')).toHaveTextContent(
      'Option 1',
    )

    await click(document.body)

    expect(screen.queryByTestId('Dropdown__option-1')).toBeNull()
  })
})
