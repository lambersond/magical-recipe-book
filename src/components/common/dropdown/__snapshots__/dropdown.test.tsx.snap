// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/common/dropdown should render correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex flex-col text-left w-fit rounded-md"
  >
    <label
      class="text-sm text-text-secondary font-bold uppercase"
      for="dropdown"
    >
      test
    </label>
    <button
      class="
          inline-flex justify-between w-72 rounded-md border border-border
          pl-4 pr-2 py-2 bg-paper text-md text-text-primary focus:outline-none
          cursor-pointer hover:bg-paper/75
        "
      data-testid="Dropdown__button"
      name="dropdown"
      type="button"
    >
      Option 1
      <svg
        class="ml-auto"
        data-testid="ArrowDownIcon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M5.23 7.21a.75.75 0 011.06.02L10 10.94l3.71-3.71a.75.75 0 111.06 1.06l-4.24 4.24a.75.75 0 01-1.06 0L5.23 8.27a.75.75 0 01.02-1.06z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`components/common/dropdown should render correctly with defaultEmpty 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex flex-col text-left w-fit rounded-md"
  >
    <label
      class="text-sm text-text-secondary font-bold uppercase"
      for="dropdown"
    >
      test
    </label>
    <button
      class="
          inline-flex justify-between w-72 rounded-md border border-border
          pl-4 pr-2 py-2 bg-paper text-md text-text-primary focus:outline-none
          cursor-pointer hover:bg-paper/75
        "
      data-testid="Dropdown__button"
      name="dropdown"
      type="button"
    >
      No options available
      <svg
        class="ml-auto"
        data-testid="ArrowDownIcon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M5.23 7.21a.75.75 0 011.06.02L10 10.94l3.71-3.71a.75.75 0 111.06 1.06l-4.24 4.24a.75.75 0 01-1.06 0L5.23 8.27a.75.75 0 01.02-1.06z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
